Select services to boot

mtojek opened the pull request:
Fixes: https://github.com/elastic/elastic-package/issues/72

Testing:
```
$ elastic-package stack up --services elasticsearch,package-registry
```

ycombinator commented:
For specifying multiple datasets as part of the `test` command we used `--dataset` (singular). Should we change that to `--datasets` (plural) to match `--services` here OR should we change `--services` to `--service` here?

elasticmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Felastic-package%2FPR-77/detail/PR-77/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Felastic-package%2FPR-77/detail/PR-77/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Felastic-package%2FPR-77/detail/PR-77/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Felastic-package%2FPR-77/detail/PR-77/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://elastic-package_77.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `[Pull request #77 opened]`

* Start Time: 2020-08-27T11:16:07.593+0000

* Duration: 2 min 15 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->





<!--COMMENT_GENERATED_WITH_ID_comment.id-->

ycombinator commented on cmd/stack.go:
```diff
@@ -10,6 +10,12 @@ import (
+var availableServices = map[string]bool{
+	"elasticsearch":    true,
+	"kibana":           true,
+	"package-registry": true,
```
Nit: Neat trick I learnt from Steffen: for such "existence lookup" maps, we can use `map[string]struct{}` as `struct{}` takes up 0 bytes whereas `bool` takes up 1 byte 😅.

mtojek replied:
You're right, let's switch to this :)

ycombinator commented on cmd/stack.go:
```diff
@@ -20,19 +26,34 @@ func setupStackCommand() *cobra.Command {
 				return cobraext.FlagParsingError(err, cobraext.DaemonModeFlagName)
 			}
 
+			services, err := cmd.Flags().GetStringSlice(cobraext.StackServicesFlagName)
```
Ah, didn't know about `GetStringSlice` - nice! Mind converting the parsing of the `--dataset` flag in the `test` command to use this same mechanism?

mtojek replied:
Fixed!

ycombinator commented on internal/stack/boot.go:
```diff
@@ -128,12 +137,18 @@ func dockerComposeUp(daemonMode bool, stackVersion string) error {
 
+	if len(options.Services) > 0 {
+		for _, aService := range options.Services {
+			args = append(args, aService, fmt.Sprintf("%s_is_ready", aService))
```
Nit: you can just append `fmt.Sprintf("%s_is_ready", aService)`. Because of the dependency, `aService` will automatically be spun up.

mtojek replied:
I can't, I checked it :) 

With single `_is_ready` service (e.g. elasticsearch_is_ready), the docker-compose will exit immediately when the "true"-application stops, so you can't achieve an effect of tailing live logs. It will behave exactly like in daemon mode.

ycombinator replied:
Ah, I see. Good to know — thanks!

mtojek commented:
> For specifying multiple datasets as part of the test command we used --dataset (singular). Should we change that to --datasets (plural) to match --services here OR should we change --services to --service here?

I renamed both, dataset and service, to plural forms.

ycombinator commented:
I was just testing this PR with `--services elasticsearch`. I ran `docker system prune -af` first to make sure I had a clean slate.

```
$ elastic-package stack up --services elasticsearch
elasticsearch uses an image, skipping
elasticsearch_is_ready uses an image, skipping
kibana uses an image, skipping
kibana_is_ready uses an image, skipping
package-registry_is_ready uses an image, skipping
Building package-registry
Step 1/3 : FROM docker.elastic.co/package-registry/distribution:snapshot
snapshot: Pulling from package-registry/distribution
524b0c1e57f8: Pull complete
e96e4b939e26: Pull complete
9815b89d68f4: Pull complete
e5885ed9db3a: Pull complete
d9caa98fa2de: Pull complete
7e79611347e7: Pull complete
516042a87580: Pull complete
61a01535d213: Pull complete
Digest: sha256:6e65e0f1dc7a6305d72814a86385ec71b8e631985c2ca48ecff4e602f441c9c1
Status: Downloaded newer image for docker.elastic.co/package-registry/distribution:snapshot
 ---> 2362c9b35f6e
Step 2/3 : COPY ${STACK_DIR}/package-registry.config.yml /package-registry/config.yml
 ---> 1f27af0c4d89
Step 3/3 : COPY ${STACK_DIR}/development/ /packages/development
 ---> 3637f360bea5

Successfully built 3637f360bea5
Successfully tagged stack_package-registry:latest
Removing network stack_default
WARNING: Network stack_default not found.
Creating network "stack_default" with the default driver
Pulling elasticsearch (docker.elastic.co/elasticsearch/elasticsearch:7.9.0)...
7.9.0: Pulling from elasticsearch/elasticsearch
f1feca467797: Pull complete
f1962334d09b: Pull complete
c3f3fd33311b: Pull complete
c01dc5464de5: Pull complete
c7c56d706a5f: Pull complete
14e364ba972c: Pull complete
8a4689df54ab: Pull complete
daa74c1eb347: Pull complete
bb1b1c773e9d: Pull complete
Digest: sha256:221c7f4a1ed3ed84ea7c1b28fd9fcb4c18e2d9370868da4b87cb78ba82f4d14e
Status: Downloaded newer image for docker.elastic.co/elasticsearch/elasticsearch:7.9.0
Pulling elasticsearch_is_ready (tianon/true:)...
latest: Pulling from tianon/true
c53fb220cbad: Pull complete
Digest: sha256:009cce421096698832595ce039aa13fa44327d96beedb84282a69d3dbcf5a81b
Status: Downloaded newer image for tianon/true:latest
Creating stack_elasticsearch_1 ... done
Creating stack_elasticsearch_is_ready_1 ... done
Attaching to stack_elasticsearch_1, stack_elasticsearch_is_ready_1
...
```

```
$ docker ps -a
CONTAINER ID        IMAGE                                                 COMMAND                  CREATED             STATUS                      PORTS                                NAMES
ca9a4209992c        tianon/true                                           "/true"                  21 seconds ago      Exited (0) 20 seconds ago                                        stack_elasticsearch_is_ready_1
2198e276b702        docker.elastic.co/elasticsearch/elasticsearch:7.9.0   "/tini -- /usr/local…"   45 seconds ago      Up 44 seconds (healthy)     127.0.0.1:9200->9200/tcp, 9300/tcp   stack_elasticsearch_1
```

```
$ docker images
REPOSITORY                                        TAG                 IMAGE ID            CREATED              SIZE
stack_package-registry                            latest              3637f360bea5        About a minute ago   342MB
docker.elastic.co/package-registry/distribution   snapshot            2362c9b35f6e        2 hours ago          342MB
docker.elastic.co/elasticsearch/elasticsearch     7.9.0               0eddd30d8bdf        2 weeks ago          762MB
tianon/true                                       latest              183cb5fd5414        22 months ago        125B
```

Notice that even though I only asked for the `elasticsearch` service to be started, the package registry image was also pulled down and the local package-registry Dockerfile was built. Is this expected? Can it be avoided unless the `package-registry` service is also supposed to be started up?

mtojek commented:
I admit I had doubts whether should I resolve it or leave it, but as you spotted it and commented, I introduced a simple service dependency adjustment for Kibana.

Fixed.

ycombinator commented on internal/cobraext/const.go:
```diff
@@ -20,6 +20,9 @@ const (
 	VerboseFlagDescription = "verbose mode"
 
+	StackServicesFlagName        = "services"
+	StackServicesFlagDescription = "component services (comma-separated values: elasticsearch, kibana, package-registry)"
```
How do you feel about introducing a `%s` placeholder instead of hardcoding the service names here and then interpolating it with `availableServices` where this constant is used?

mtojek replied:
Fixed

ycombinator commented on cmd/stack.go:
```diff
@@ -90,3 +115,28 @@ func setupStackCommand() *cobra.Command {
 }
+
+func availableServicesAsList() []string {
+	var available []string
```
Nit: Not that it matters much here because we are only dealing with a handful of `availableServices` but in general, when the size of the slice can be determined beforehand you can preallocate the memory for it by doing:

```
available := make([]string, len(availableServices)
```

And then you don't even need the `append` because you can do something like:

```
i := 0
for aService := range availableServices {
  available[i] = aService
  i++
}
```

But again, this is just to show a _general_ pattern when converting the keys of a `map[string]interface{}` to a `[]string`. What you have here is totally fine since we are only dealing with a handful of keys.

mtojek replied:
You're right, we can optimize it here :)

Fixed.
